037-prototypes/proposal.md

```diff
@@ -0,0 +1,473 @@
+
+One example of a message that will yield new objects is `check` (part of the
+[resource interface](https://github.com/concourse/rfcs/pull/26)). This message
+will be sent to prototypes that are used as a resource in a Concourse pipeline.
+
+Per the resource interface, the `check` message returns an object for each
+version. However, these objects aren't self-contained - instead, they are an
+implicit *clone* of the original object, with the returned object's fields
+merged in.
+
+For example, the `git` prototype may be initially receive the following
+user-configured properties in a `check` message:
+
+```json
+{
+  "object": {
+    "uri": "https://github.com/concourse/rfcs",
+    "branch": "master"
+  }
+}
+```
+
+Let's say the `check` handler responds with the following objects, representing
+two commits in the repository (with `metadata` omitted for brevity):
+
+```json
+{
+  "object": {
+    "ref": "e4be0b367d7bd34580f4842dd09e7b59b6097b25"
+  }
+}
+{
+  "object": {
+    "ref": "5a052ba6438d754f73252283c6b6429f2a74dbff"
+  }
+}
+```
+
+Per the resource interface, these object properties would be saved as versions
+of the resource in the order returned by `check`.
+
+When it comes time to fetch a version in a build, the `git` prototype would
+receive a `get` message with the following object:
+
+```json
+{
+  "object": {
+    "uri": "https://github.com/concourse/rfcs",
+    "branch": "master",
+    "ref": "e4be0b367d7bd34580f4842dd09e7b59b6097b25"
+  }
+}
+```
+
+Note that the object properties have been merged into the original set of
+properties.
+
+## Pipeline Usage
+
+A new `run` step type will be introduced. It takes the name of a message to
+send, and a `type` denoting the prototype to use:
+
+```yaml
+prototypes:
+- name: some-prototype
+  type: registry-image
+  source:
+    repository: vito/some-prototype
+
+jobs:
+- name: run-prototype
+  plan:
+  - run: some-message
+    type: some-prototype
+```
+
+So far this example is not very useful. There aren't many things I can think to
+run that don't take any inputs or configuration or produce any outputs. Let's
+try giving it an input and collecting its resulting output. This can be done
+with `inputs`/`input_mapping` and `outputs`/`output_mapping`:
+
+```yaml
+prototypes:
+- name: oci-image
+  type: registry-image
+  source:
+    repository: vito/oci-image-prototype
+
+resources:
+- name: my-image-src
+  type: git
+  source:
+    uri: https://example.com/my-image-src
+
+- name: my-image
+  type: registry-image
+  source:
+    repository: example/my-image
+    username: some-username
+    password: some-password
+
+jobs:
+- name: build-and-push
+  plan:
+  - get: my-image-src
+  - run: build
+    type: oci-image
+    input_mapping: {context: my-image-src}
+    outputs: [image]
+  - put: my-image
+    params: {image: image/image.tar}
+```
+
+This example also makes use of resources, which are also implemented as
+prototypes with special pipeline pipeline semantics and step syntax. These
```

Comment:
"Special pipeline semantics" really, the invariants I reference below codify a relationship between the high level "pipeline" concept and "resource" prototypes, which together define a category of sorts with desirable structural properties. My open question is, "does affordance for other categories really solve problems for people?" An important part of resources today is that every resource implements the same prototype, and *imply* the abstraction "pipeline" (which is the thing I really want). So while I can imagine an oci image prototype existing, and I can imagine how it "would work" i don't understand what it would "be for" without a complementary high level object that it *implies*.

Reply:
> My open question is, "does affordance for other categories really solve problems for people?"

Yes. These problems are cited as motivations for this PR and other PRs which describe other use cases for prototypes:

* https://github.com/vito/rfcs/blob/spatial-resources/029-across-step/proposal.md#motivation
* https://github.com/vito/rfcs/blob/notification-resources/028-notification-resources/proposal.md#motivation

The magnitude of these proposals really cannot be overstated. It is the core focus of the v10 roadmap which is based on years of user feedback and complaints over not properly supporting pull requests and notifications.

Beyond these two concrete workflows, the need for a better way to distribute arbitrary functionality has been brought up since the very early days of Concourse, and is typically cited as concourse/rfcs#7.

There is also a path forward to finally solving https://github.com/concourse/concourse/issues/3023 through the use of prototypes for [https://github.com/vito/rfcs/blob/var-sources/039-var-sources/proposal.md), which would allow credentials to be fetched by using a prototype on a worker with IAM roles configured. This has also been a user need for years.

Further, prototypes can be used to represent things which should trigger builds, but are not necessarily *versioned artifacts* (which imply a central, globally verifiable artifact history), thereby solving the "thundering herd" problem of having many jobs downstream of a 10 minute interval, using the [`get_var` step](https://github.com/vito/rfcs/blob/trigger-resources/027-var-steps/proposal.md).